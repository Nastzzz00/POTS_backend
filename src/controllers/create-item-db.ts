import { IDBModel } from '../commons/types';
import { createUCCreateItem } from '../core/';

const createCreateItemDB = (db: IDBModel<any>) => async i => {
  const createItem = createUCCreateItem();
  const newItem = createItem(i);
  return db.insert(newItem);
};

export { createCreateItemDB };
