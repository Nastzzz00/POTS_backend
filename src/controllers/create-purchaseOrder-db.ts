import { IDBModel } from '../commons/types';
import { createUCPurchaseOrder } from '../core/';

const createCreatePurchaseOrderDB = (db: IDBModel<any>) => async po => {
	const createPurchaseOrder = createUCPurchaseOrder();
	const newPurchaseOrder = createPurchaseOrder(po);
	return db.insert(newPurchaseOrder);
};

export { createCreatePurchaseOrderDB };
