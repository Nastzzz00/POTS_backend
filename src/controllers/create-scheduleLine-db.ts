import { IDBModel } from '../commons/types';
import { createUCCreateScheduleLine } from '../core/';

const createCreateScheduleLineDB = (db: IDBModel<any>) => async sl => {
	const createScheduleLine = createUCCreateScheduleLine();
	const newScheduleLine = createScheduleLine(sl);
	return db.insert(newScheduleLine);
};

export { createCreateScheduleLineDB };
