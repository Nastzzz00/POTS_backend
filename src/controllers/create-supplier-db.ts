import { IDBModel } from '../commons/types';
import { createUCCreateSupplier } from '../core/';

const createCreateSupplierDB = (db: IDBModel<any>) => async u => {
	const createSupplier = createUCCreateSupplier();
	const newSupplier = createSupplier(u);
	return db.insert(newSupplier);
};

export { createCreateSupplierDB };
