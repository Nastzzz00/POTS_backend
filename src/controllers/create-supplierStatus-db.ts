import { IDBModel } from '../commons/types';
import { createUCCreateSupplierStatus } from '../core/';

const createCreateSupplierStatusDB = (db: IDBModel<any>) => async u => {
	const createSupplierStatus = createUCCreateSupplierStatus();
	const newSupplierStatus = createSupplierStatus(u);
	return db.insert(newSupplierStatus);
};

export { createCreateSupplierStatusDB };
