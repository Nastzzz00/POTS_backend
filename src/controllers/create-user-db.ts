import { IDBModel } from '../commons/types';
import { createUCCreateUser } from '../core/';

const createCreateUserDB = (db: IDBModel<any>) => async u => {
	const createUser = createUCCreateUser();
	const newUser = createUser(u);
	return db.insert(newUser);
};

export { createCreateUserDB };
