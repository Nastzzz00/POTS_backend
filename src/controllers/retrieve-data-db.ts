import { IDBModel } from '../commons/types';

const getAllDataDB = (db: IDBModel<any>) => async () => {
	return db.getAll();
};
const getByIDDB = (db: IDBModel<any>) => async id => {
	return db.getById(id);
};

const getAllBySupplierStatusDB = (db: IDBModel<any>) => async id => {
	return db.getAllBySupplierStatus(id);
};

const getAllByItemDB = (db: IDBModel<any>) => async id => {
	return db.getAllByItem(id);
};

const getAllByScheduleLineDB = (db: IDBModel<any>) => async id => {
	return db.getAllByScheduleLine(id);
};

export { getAllDataDB, getByIDDB, getAllByItemDB, getAllBySupplierStatusDB, getAllByScheduleLineDB };
