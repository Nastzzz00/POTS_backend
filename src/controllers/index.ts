import {
  getAllDataDB,
  getByIDDB,
  getAllByItemDB,
  getAllBySupplierStatusDB,
  getAllByScheduleLineDB,
} from './retrieve-data-db';
import { DeleteRecordByIDDB } from './delete-data-db';

//User
import { createCreateUserDB } from './create-user-db';
import { updateUserByIDDB } from './update-user-db';

//Supplier
import { createCreateSupplierDB } from './create-supplier-db';
import { updateSupplierByIDDB } from './update-supplier-db';

//Supplier Status
import { createCreateSupplierStatusDB } from './create-supplierStatus-db';
import { updateSupplierStatusByIDDB } from './update-supplierStatus-db';
import { updateSupplierStatusItemByIDDB } from './update-supplierStatusItem-db';

//Item
import { createCreateItemDB } from './create-item-db';
import { updateItemByIDDB } from './update-item-db';

//Purchase Order
import { createCreatePurchaseOrderDB } from './create-purchaseOrder-db';
import { updatePurchaseOrderByIDDB } from './update-purchaseOrder-db';
import { updateAdminStatusByIDDB } from './update-adminStatus-db';

//Schedule Line
import { createCreateScheduleLineDB } from './create-scheduleLine-db';
import { updateScheduleLineByIDDB } from './update-scheduleLine-db';

export {
  getAllDataDB,
  getByIDDB,
  DeleteRecordByIDDB,
  getAllByItemDB,
  getAllBySupplierStatusDB,
  getAllByScheduleLineDB,
  //User exports
  createCreateUserDB,
  updateUserByIDDB,
  //Supplier exports
  createCreateSupplierDB,
  updateSupplierByIDDB,
  //SupplierStatus exports
  createCreateSupplierStatusDB,
  updateSupplierStatusByIDDB,
  //Item
  createCreateItemDB,
  updateItemByIDDB,
  updateSupplierStatusItemByIDDB,
  //Purchase Order
  createCreatePurchaseOrderDB,
  updatePurchaseOrderByIDDB,
  updateAdminStatusByIDDB,
  //Schedule Line
  createCreateScheduleLineDB,
  updateScheduleLineByIDDB,
};
