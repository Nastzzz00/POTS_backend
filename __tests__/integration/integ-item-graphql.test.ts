import { createTestClient } from 'apollo-server-testing';
import gql from 'graphql-tag';
import mockData from '../mocks/mock-data';
import * as controllers from '../../src/controllers';
import { constructTestServer } from '../__utils';

const {
  getAllDataDB,
  getByIDDB,
  getAllBySupplierStatusDB,
  getAllByScheduleLineDB,
  DeleteRecordByIDDB,
  createCreateItemDB,
  updateItemByIDDB,
  updateSupplierStatusItemByIDDB,
  createCreateScheduleLineDB,
  createCreateSupplierStatusDB,
} = controllers;

const itemMock = {
  insert: jest.fn(async input => {
    return {
      id: '1',
      ...input,
      deliveryAddress: 'A1',
    };
  }),
  getById: jest.fn(async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.items.filter(filterData);

    return res[0] || null;
  }),
  getAll: jest.fn(async () => {
    return mockData.items;
  }),
  updateById: jest.fn(async input => {
    return { ...input };
  }),
  deleteById: async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.items.filter(filterData);
    return res[0] || null;
  },
  getAllBySupplierStatus: async id => {},
  getAllByItem: async id => {},
  getAllByScheduleLine: jest.fn(async id => {
    const filterData = data => {
      if (id.includes(data._id)) {
        return data;
      }
    };
    const res = mockData.scheduleLines
      .filter(filterData)
      .map(data => ({ ...data, id: data._id }));
    return res;
  }),
  updateSupplierStatusItemById: async input => {
    return { ...input };
  },
  updateAdminStatusPurchaseOrderById: async id => {},
};

const supplierStatusMock = {
  insert: jest.fn(async input => {
    return { id: '1', ...input };
  }),
  getById: jest.fn(async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.supplierStatus.filter(filterData);

    return res[0] || null;
  }),
  getAll: jest.fn(async () => {
    return mockData.supplierStatus;
  }),
  updateById: jest.fn(async input => {
    return { ...input };
  }),
  deleteById: async id => {},
  getAllBySupplierStatus: async id => {},
  getAllByItem: async id => {},
  getAllByScheduleLine: async id => {},
  updateSupplierStatusItemById: async id => {},
  updateAdminStatusPurchaseOrderById: async id => {},
};

const scheduleLinesMock = {
  insert: jest.fn(async input => {
    return { id: '2', ...input };
  }),
  getById: jest.fn(async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.scheduleLines.filter(filterData);

    return res[0] || null;
  }),
  getAll: jest.fn(async () => {
    return mockData.scheduleLines;
  }),
  updateById: jest.fn(async input => {
    return { ...input };
  }),
  deleteById: async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.scheduleLines.filter(filterData);
    return res[0] || null;
  },
  getAllBySupplierStatus: jest.fn(async id => {
    const filterData = data => {
      if (id.includes(data._id)) {
        return data;
      }
    };
    const res = mockData.supplierStatus
      .filter(filterData)
      .map(data => ({ ...data, id: data._id }));
    return res;
  }),
  getAllByItem: async id => {},
  getAllByScheduleLine: async id => {},
  updateSupplierStatusItemById: async id => {},
  updateAdminStatusPurchaseOrderById: async id => {},
};

const addressMock = {
  insert: jest.fn(async input => {
    return { id: 'A1', ...input };
  }),
  getAll: jest.fn(async () => {
    return mockData.address;
  }),
  getById: jest.fn(async id => {
    const filterData = data => {
      if (data.id === id) {
        return data;
      }
    };
    const res = mockData.address.filter(filterData);
    return res[0] || null;
  }),
  deleteById: async id => {},
  updateById: async id => {},
  getAllBySupplierStatus: async id => {},
  getAllByItem: async id => {},
  getAllByScheduleLine: async id => {},
  updateSupplierStatusItemById: async id => {},
  updateAdminStatusPurchaseOrderById: async id => {},
};

const { server }: any = constructTestServer({
  context: {
    createItem: createCreateItemDB(itemMock),
    updateItemById: updateItemByIDDB(itemMock),
    updateItemSupplierStatusById: updateSupplierStatusItemByIDDB(itemMock),
    deleteItemById: DeleteRecordByIDDB(itemMock),
    getItemById: getByIDDB(itemMock),
    getAllItems: getAllDataDB(itemMock),
    getAllSupplierStatusByScheduleLine: getAllBySupplierStatusDB(
      scheduleLinesMock,
    ),
    getAllScheduleLinesByItem: getAllByScheduleLineDB(itemMock),
    createScheduleLine: createCreateScheduleLineDB(scheduleLinesMock),
    getAddressById: getByIDDB(addressMock),
    createSupplierStatus: createCreateSupplierStatusDB(supplierStatusMock),
  },
});

describe('Tests', () => {
  //Queries
  it('should fetch all items', async () => {
    const ITEM_ALL = gql`
      query {
        allItems {
          id
          itemNo
          description
          quantity
          uom
          unitPrice
          currency
          deliveryAddress {
            id
          }
          totalAmount
          supplierStatusItem
          scheduleLine {
            id
            quantity
            uom
            unitPrice
            totalAmount
            deliveryDateAndTime
            deliveryStatus {
              id
              status
              dateCreated
              timeCreated
            }
          }
        }
      }
    `;

    const { query } = createTestClient(server);
    const res = await query({ query: ITEM_ALL });

    expect(res).toMatchSnapshot();
  });

  it('should fetch one item ', async () => {
    const SINGLE_ITEM = gql`
      query i($id: String!) {
        item(id: $id) {
          id
          itemNo
          description
          quantity
          uom
          unitPrice
          currency
          deliveryAddress {
            id
          }
          totalAmount
          supplierStatusItem
          scheduleLine {
            id
            quantity
            uom
            unitPrice
            totalAmount
            deliveryDateAndTime
            deliveryStatus {
              id
              status
              dateCreated
              timeCreated
            }
          }
        }
      }
    `;

    const { query } = createTestClient(server);
    const res = await query({
      query: SINGLE_ITEM,
      variables: { id: '1' },
    });

    expect(res).toMatchSnapshot();
  });

  it('should error when no item', async () => {
    const SINGLE_ITEM = gql`
      query i($id: String!) {
        item(id: $id) {
          id
          itemNo
          description
          quantity
          uom
          unitPrice
          currency
          deliveryAddress {
            id
          }
          totalAmount
          supplierStatusItem
          scheduleLine {
            id
            quantity
            uom
            unitPrice
            totalAmount
            deliveryDateAndTime
            deliveryStatus {
              id
              status
              dateCreated
              timeCreated
            }
          }
        }
      }
    `;

    const { query } = createTestClient(server);
    const res = await query({
      query: SINGLE_ITEM,
      variables: {
        id: '',
      },
    });

    expect(res).toMatchSnapshot();
  });

  //Mutations

  it('should create an item', async () => {
    const CREATE_ITEM = gql`
      mutation createI($item: ItemInput!) {
        createItem(item: $item) {
          id
          itemNo
          productId
          description
          quantity
          uom
          unitPrice
          discount
          totalAmount
          deliveryAddress {
            building_name
            zip_code
            state
            street
            city
          }
          scheduleLine {
            id
            deliveryDateAndTime
            quantity
            uom
            unitPrice
            totalAmount
            deliveryStatus {
              id
              status
              dateCreated
              timeCreated
            }
          }
          currency
        }
      }
    `;

    const { mutate } = createTestClient(server);
    const res = await mutate({
      mutation: CREATE_ITEM,
      variables: {
        item: {
          itemNo: '001',
          description: 'Beef',
          productId: '0001',
          quantity: 10,
          uom: 'kilograms',
          unitPrice: 1000,
          discount: 0.05,
          totalAmount: 10000,
          deliveryAddress: {
            building_name: '002',
            street: 'Elmer',
            city: 'Celadon',
            state: 'Johto',
            zip_code: '123',
          },
          scheduleLine: [
            {
              quantity: 10,
              uom: 'kilograms',
              deliveryDateAndTime: 'February 25,2020 4:30PM',
              unitPrice: 1000,
              totalAmount: 10000,
              deliveryStatus: [
                {
                  status: 'Delivered',
                },
              ],
            },
          ],
          currency: 'PHP',
        },
      },
    });

    expect(res.errors).toBeUndefined();
    expect(itemMock.insert.mock.calls.length).toBe(1);

    expect(res.data).toMatchObject({
      createItem: {
        id: '1',
        itemNo: '001',
        description: 'Beef',
        productId: '0001',
        quantity: 10,
        uom: 'kilograms',
        unitPrice: 1000,
        discount: 0.05,
        totalAmount: 10000,
        deliveryAddress: {
          building_name: '002',
          street: 'Elmer',
          city: 'Celadon',
          state: 'Johto',
          zip_code: '123',
        },
        scheduleLine: [
          {
            id: '2',
            quantity: 10,
            uom: 'kilograms',
            deliveryDateAndTime: 'February 25,2020 4:30PM',
            unitPrice: 1000,
            totalAmount: 10000,
            deliveryStatus: [
              {
                id: '2',
                status: 'Delivered',
                dateCreated: 'February 14, 2020',
                timeCreated: '4:30 PM',
              },
            ],
          },
        ],
        currency: 'PHP',
      },
    });
    expect(res).toMatchSnapshot();
  });

  it('update an item', async () => {
    const UPDATE_ITEM = gql`
      mutation i($item: UpdateItemInput!) {
        updateItem(item: $item) {
          id
          supplierStatusItem
        }
      }
    `;

    const { mutate } = createTestClient(server);
    const res = await mutate({
      mutation: UPDATE_ITEM,
      variables: {
        item: {
          id: '1',
          supplierStatusItem: 'Dispatched',
        },
      },
    });

    expect(res).toMatchSnapshot();
  });

  it('update an item supplierStatus', async () => {
    const UPDATE_ITEM = gql`
      mutation i($item: UpdateSupplierStatusItemInput!) {
        updateSupplierStatusItem(item: $item) {
          id
          supplierStatusItem
        }
      }
    `;

    const { mutate } = createTestClient(server);
    const res = await mutate({
      mutation: UPDATE_ITEM,
      variables: {
        item: {
          id: '1',
          supplierStatusItem: 'Dispatched',
        },
      },
    });

    expect(res).toMatchSnapshot();
  });

  it('delete an item', async () => {
    const DELETE_ITEM = gql`
      mutation i($id: ID!) {
        deleteItem(id: $id) {
          id
        }
      }
    `;

    const { mutate } = createTestClient(server);
    const res = await mutate({
      mutation: DELETE_ITEM,
      variables: { id: '1' },
    });

    expect(res).toMatchSnapshot();
  });
});
