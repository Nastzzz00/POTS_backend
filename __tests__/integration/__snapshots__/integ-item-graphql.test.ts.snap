// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests delete an item 1`] = `
Object {
  "data": Object {
    "deleteItem": Object {
      "id": "1",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests should create an item 1`] = `
Object {
  "data": Object {
    "createItem": Object {
      "currency": "PHP",
      "deliveryAddress": Object {
        "building_name": "002",
        "city": "Celadon",
        "state": "Johto",
        "street": "Elmer",
        "zip_code": "123",
      },
      "description": "Beef",
      "discount": 0.05,
      "id": "1",
      "itemNo": "001",
      "productId": "0001",
      "quantity": 10,
      "scheduleLine": Array [
        Object {
          "deliveryDateAndTime": "February 25,2020 4:30PM",
          "deliveryStatus": Array [
            Object {
              "dateCreated": "February 14, 2020",
              "id": "2",
              "status": "Delivered",
              "timeCreated": "4:30 PM",
            },
          ],
          "id": "2",
          "quantity": 10,
          "totalAmount": 10000,
          "unitPrice": 1000,
          "uom": "kilograms",
        },
      ],
      "totalAmount": 10000,
      "unitPrice": 1000,
      "uom": "kilograms",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests should error when no item 1`] = `
Object {
  "data": Object {
    "item": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests should fetch all items 1`] = `
Object {
  "data": Object {
    "allItems": Array [
      Object {
        "currency": "PHP",
        "deliveryAddress": Object {
          "id": "A1",
        },
        "description": "Beef",
        "id": "1",
        "itemNo": "001",
        "quantity": 10,
        "scheduleLine": Array [
          Object {
            "deliveryDateAndTime": "February 25,2020 4:30PM",
            "deliveryStatus": Array [],
            "id": "1",
            "quantity": 10,
            "totalAmount": 10000,
            "unitPrice": 1000,
            "uom": "kilograms",
          },
        ],
        "supplierStatusItem": null,
        "totalAmount": 10000,
        "unitPrice": 1000,
        "uom": "kilograms",
      },
      Object {
        "currency": "PHP",
        "deliveryAddress": Object {
          "id": "A1",
        },
        "description": "Beef",
        "id": "2",
        "itemNo": "001",
        "quantity": 10,
        "scheduleLine": Array [
          Object {
            "deliveryDateAndTime": "February 25,2020 4:30PM",
            "deliveryStatus": Array [
              Object {
                "dateCreated": "February 14, 2020",
                "id": "2",
                "status": "Delivered",
                "timeCreated": "4:30 PM",
              },
            ],
            "id": "2",
            "quantity": 10,
            "totalAmount": 10000,
            "unitPrice": 1000,
            "uom": "kilograms",
          },
        ],
        "supplierStatusItem": null,
        "totalAmount": 10000,
        "unitPrice": 1000,
        "uom": "kilograms",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests should fetch one item  1`] = `
Object {
  "data": Object {
    "item": Object {
      "currency": "PHP",
      "deliveryAddress": Object {
        "id": "A1",
      },
      "description": "Beef",
      "id": "1",
      "itemNo": "001",
      "quantity": 10,
      "scheduleLine": Array [
        Object {
          "deliveryDateAndTime": "February 25,2020 4:30PM",
          "deliveryStatus": Array [],
          "id": "1",
          "quantity": 10,
          "totalAmount": 10000,
          "unitPrice": 1000,
          "uom": "kilograms",
        },
      ],
      "supplierStatusItem": null,
      "totalAmount": 10000,
      "unitPrice": 1000,
      "uom": "kilograms",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests update an item 1`] = `
Object {
  "data": Object {
    "updateItem": Object {
      "id": "1",
      "supplierStatusItem": "Dispatched",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`Tests update an item supplierStatus 1`] = `
Object {
  "data": Object {
    "updateSupplierStatusItem": null,
  },
  "errors": Array [
    [GraphQLError: updateSupplierStatusItemById is not a function],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
